### PR TITLE
feat(KIT-0036): manifest sync engine + workflow refactor (PR 1 of 2)

### PR DESCRIPTION
**Pull request by movito**
## Summary

PR 1 of 2 implementing **KIT-ADR-0026** (pull-based consumer sync). Extracts Channel B's sync logic from the `sync-core-scripts.yml` heredoc into a single tested Python engine — **one engine, two callers** — so the push Action and (in PR 2) the consumer-side `project sync` command cannot drift.

This PR delivers the engine, tests, workflow refactor, and `workflow_dispatch`. PR 2 adds the `project sync` wrapper, retires `check-sync.sh`, and bumps to v3.0.0.

## What's here (D1–D4)

- **D1 `scripts/core/sync_from_manifest.py`** — library-first `sync(source, target, options) -> SyncReport`; thin argparse CLI mapping `SyncReport.status` to the exit-code contract. Faithful port of the tier/path/opt-in/cleanup/overwrite-warning logic. `--tier` / `--only` narrowing, `--dry-run`, `--report-json`. Two-pass temp-then-commit (KIT-0034 F3), full plan (bytes **and** mode) read into memory before the first write → safe self-update, exec bits preserved. Engine-computed completeness (`partial_sync` marker).
- **D2 `tests/test_sync_from_manifest.py`** — 39 tests, **96% coverage**: characterization (tree-property, not gold-file), dual-entrypoint contract (library vs CLI), self-sync (in-process **and** subprocess proving the running engine can overwrite itself), frozen exit-code contract, spec edge cases.
- **D2b manifest** — add `core/sync_from_manifest.py` to `scripts_core`; bump `core_version` + `VERSION` 2.1.0 → **2.2.0**; entry-count tests updated (18 / 43).
- **D3 workflow** — the sync step is now a thin shell around the engine; behavior gated by the characterization tests; fails only on engine exit ≥ 2.
- **D4 `workflow_dispatch`** — optional `repo` input to sync one downstream repo on demand.

## Exit-code contract (frozen by tests)

`0` clean/applied · `1` drift (`--dry-run`) or applied-with-warnings · `2` usage · `3` source manifest missing/unrecognized · `4` source unreadable.

## Test plan

- [x] `pytest` green (328 → 367 tests; 39 new)
- [x] New-code coverage 96% (≥ 80% gate)
- [x] `pattern_lint.py` clean; black / isort / flake8 / DK rules pass
- [x] Engine run against a local `--source` fixture completes in well under 5 s
- [x] `--tier` pull sets `partial_sync: true`; a following full pull clears it and bumps `core_version`
- [x] Workflow YAML validates

## Notes for PR 2

- `project` already aliases `sync` → `linearsync`; PR 2 must resolve that collision for `project sync`.
- The dual-entrypoint test becomes a true dual-**caller** test once `project sync` lands.

Design authority: `.kit/adr/KIT-ADR-0026-pull-based-consumer-sync.md`
Task spec: `.kit/tasks/3-in-progress/KIT-0036-pull-based-consumer-sync.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes how vendored files are written into downstream repos via GitHub Actions and a new sync engine; mistakes could corrupt consumer trees or open bad sync PRs, though characterization tests and workflow exit-code gating mitigate this.
> 
> **Overview**
> Implements **KIT-ADR-0026** PR 1: Channel B manifest sync moves out of an untested ~110-line bash heredoc in `sync-core-scripts.yml` into **`scripts/core/sync_from_manifest.py`** — a library-first engine (`sync()` → `SyncReport`) with a thin CLI, frozen exit codes, two-pass temp-then-commit apply, tier/opt-in/`partial_sync` semantics, and `--dry-run` / `--tier` / `--only` / `--report-json`.
> 
> The push workflow is now a thin shell that runs the engine, surfaces warnings via JSON, stages only manifest roots (not `git add -A`), and uses a randomized multiline `GITHUB_OUTPUT` delimiter. **Manual `workflow_dispatch`** adds an optional `repo` filter, a **`validate-dispatch`** job so typos fail loudly, and **per-repo concurrency** to avoid races.
> 
> **`core_version` / `VERSION` bump to 2.2.0** with the engine listed in `scripts_core`; **`tests/test_sync_from_manifest.py`** (~52 tests) gates behavior. Consumer **`project sync`** and `check-sync.sh` retirement land in PR 2.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0d5f35c63333f9cdb78949609284f518e72fa09d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/core/sync_from_manifest.py [hunk 1]
```diff
@@ -0,0 +1,745 @@
...
+    manifest_changed = _manifest_without_synced_at(
+        new_manifest
+    ) != _manifest_without_synced_at(target_manifest)
+    should_write = content_changed or manifest_changed
+
+    synced_tiers = sorted({tier for tier, _entry in candidates})
+
+    # ── Status determination ──
+    if options.dry_run:
+        status = "drift" if content_changed else "clean"
```
[the remark this answers is not in the document]
Comment (movito) [reply scripts/core/sync_from_manifest.py]: Fixed in 7473e9a. Dry-run status now mirrors `should_write` (content change **or** manifest-metadata change), so a `core_version` bump or `partial_sync` clear with identical file content reports `drift` (exit 1). `synced_at` alone still never counts as drift. Added `test_dry_run_flags_version_only_drift`.

File: scripts/core/sync_from_manifest.py [hunk 2]
```diff
@@ -0,0 +1,745 @@
...
+    consumer has none, and a consumer on an older schema should still get its
+    ``opted_in`` and prior entry set read where possible.
+    """
+    manifest_path = target_root / MANIFEST_RELPATH
+    if not manifest_path.is_file():
+        return None
+    try:
+        data = _read_json(manifest_path)
+    except (json.JSONDecodeError, UnicodeDecodeError, OSError):
+        return None
```
[the remark this answers is not in the document]
Comment (movito) [reply scripts/core/sync_from_manifest.py]: Fixed in 7473e9a. A target manifest that exists but is corrupt (unparseable, or not a JSON object) now raises `ManifestError` (exit 3) instead of being treated as fresh — the old path silently reset `opted_in` to `[]`. An absent manifest still means a genuinely fresh consumer; an older-but-valid dict is still read leniently. Added `test_corrupt_target_manifest_raises_not_silent_reset` and `test_non_object_target_manifest_raises`.

File: .github/workflows/sync-core-scripts.yml [hunk 3]
```diff
@@ -21,6 +21,14 @@ on:
...
+    inputs:
+      repo:
+        description: >-
+          Sync only this downstream repo (owner/name). Leave blank to sync
+          every repo in the matrix.
+        required: false
+        default: ''
 
 jobs:
   sync:
```
[the remark this answers is not in the document]
Comment (movito) [reply .github/workflows/sync-core-scripts.yml:53]: Accepted — fixed in d12a627. Added a per-repo `concurrency` group (`sync-core-scripts-${{ matrix.repo }}`, `cancel-in-progress: false`) so a push sync and a workflow_dispatch can't race on the same target. This is exactly the race the ADR flags in Consequences.

File: .github/workflows/sync-core-scripts.yml [hunk 4]
```diff
@@ -36,11 +44,19 @@ jobs:
             is_kit: true
     steps:
       - name: Checkout source (agentive-starter-kit)
+        if: >-
+          github.event_name == 'push'
+          || github.event.inputs.repo == ''
+          || github.event.inputs.repo == matrix.repo
         uses: actions/checkout@v6
         with:
           path: source
```
[the remark this answers is not in the document]
Comment (movito) [reply .github/workflows/sync-core-scripts.yml:84]: Declining for this PR (with reason): `actions/checkout@v6` is unpinned throughout this workflow already (pre-existing), and pinning to SHAs is a repo-wide supply-chain policy decision that should be applied consistently across all workflows in one pass — out of scope for KIT-0036, which is about sync logic. Worth a dedicated hardening task.

File: .github/workflows/sync-core-scripts.yml [hunk 5]
```diff
@@ -49,147 +65,72 @@ jobs:
...
           if [ -n "$WARNINGS" ]; then
-            echo "warnings<<EOF" >> "$GITHUB_OUTPUT"
-            echo -e "$WARNINGS" >> "$GITHUB_OUTPUT"
-            echo "EOF" >> "$GITHUB_OUTPUT"
+            {
+              echo "warnings<<EOF"
+              echo "$WARNINGS"
+              echo "EOF"
+            } >> "$GITHUB_OUTPUT"
           fi
```
[the remark this answers is not in the document]
Comment (movito) [reply .github/workflows/sync-core-scripts.yml:166]: Accepted (Major) — fixed in d12a627. The multiline `GITHUB_OUTPUT` write now uses a random delimiter (`EOF_$(openssl rand -hex 16)`) instead of a fixed `EOF`, closing the env-file injection vector on the engine-derived WARNINGS value.

File: .kit/adr/KIT-ADR-0026-pull-based-consumer-sync.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
+# KIT-ADR-0026: Pull-Based Consumer Sync — One Engine, Two Callers
+
+**Status**: Proposed
+**Date**: 2026-07-03 (amended same day after adversarial review — see
+Evaluation trail)
```
[the remark this answers is not in the document]
Comment (movito) [reply .kit/adr/KIT-ADR-0026-pull-based-consumer-sync.md]: Accepted — fixed in d12a627. ADR-0026 status is now `Accepted`, matching the evaluation trail (two review rounds, final APPROVED).

File: .kit/context/KIT-0036-HANDOFF-2026-07-03.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
+# KIT-0036 Handoff — Pull-based consumer sync (engine + `project sync`)
+
+**Date**: 2026-07-03
+**From**: planning session "2026-07-03 FDF5 Better distribution setups"
+**Task spec**: `.kit/tasks/2-todo/KIT-0036-pull-based-consumer-sync.md`
+**Design authority**: `.kit/adr/KIT-ADR-0026-pull-based-consumer-sync.md`
```
[the remark this answers is not in the document]
Comment (movito) [reply .kit/context/KIT-0036-HANDOFF-2026-07-03.md]: Accepted — fixed in d12a627. The handoff task-spec link now points at `3-in-progress` (the task moved when work started).

File: scripts/core/sync_from_manifest.py [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+            for rel, content in op.dir_files.items():
+                full = f"{op.relpath}/{rel}"
+                if rel not in existing:
+                    added.append(full)
+                elif existing[rel].data != content.data:
+                    modified.append(full)
+            for rel in existing:
+                if rel not in op.dir_files:
+                    removed_files.append(f"{op.relpath}/{rel}")
+    return added, modified, removed_files, warnings
```
[the remark this answers is not in the document]
Comment (movito) [reply scripts/core/sync_from_manifest.py]: Accepted (good catch) — fixed in d12a627. `_diff_plan` now compares file `mode` as well as bytes for both file and directory entries, so a file that only lost its exec bit is real drift (`_apply` restores mode via chmod). Added `test_mode_only_change_is_drift`.

File: scripts/core/sync_from_manifest.py [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+        status = "drift" if content_changed else "clean"
+    else:
+        if should_write:
+            _apply(plan, new_manifest, target_root)
+        if warnings:
+            status = "applied_with_warnings"
+        elif content_changed:
+            status = "applied"
+        else:
+            status = "clean"
```
[the remark this answers is not in the document]
Comment (movito) [reply scripts/core/sync_from_manifest.py]: Already addressed in 7473e9a (as you detected). Dry-run status now mirrors `should_write`, so a manifest-only change (core_version bump, partial_sync clear) is reported as drift.

File: tests/test_sync_from_manifest.py [hunk 10]
```diff
@@ -0,0 +1,531 @@
...
+        synced = (target / "scripts" / "core" / "sync_from_manifest.py").read_text(
+            encoding="utf-8"
+        )
+        assert synced == modified_engine
+
+
+def read_manifest_source(root: Path) -> dict:
+    with open(root / "scripts" / ".core-manifest.json", encoding="utf-8") as handle:
+        return json.load(handle)
+
```
[the remark this answers is not in the document]
Comment (movito) [reply tests/test_sync_from_manifest.py]: Accepted — fixed in d12a627. Dropped `read_manifest_source`; the self-sync tests now use the module-level `read_manifest` helper.

File: .github/workflows/sync-core-scripts.yml [hunk 11]
```diff
@@ -49,147 +65,72 @@ jobs:
...
 
-          git add scripts/core/ scripts/.core-manifest.json
-          if [ -d .claude/commands/ ]; then
-            git add .claude/commands/
-          fi
-          if [ -d .kit/ ]; then
-            git add .kit/
-          fi
+          git add -A
```
[the remark this answers is not in the document]
Comment (movito) [reply .github/workflows/sync-core-scripts.yml]: Fixed in ab071b8. Replaced `git add -A` with scoped adds of only the roots the manifest targets (`scripts/`, `.claude/commands/`, `.kit/`, `.adversarial/`). Bonus: the previous scoped adds never staged `.adversarial/`, which the kit_builder tier writes — so this also fixes a latent omission.

File: .github/workflows/sync-core-scripts.yml [hunk 12]
```diff
@@ -222,12 +163,17 @@ jobs:
 
           The \`kit_builder\` tier syncs automatically to kit repos and does not require opt-in.
 
+          ## Pull-based alternative
+
+          You can also sync on demand from your own checkout with
+          \`./scripts/core/project sync --dry-run\` (see KIT-ADR-0026).
```
[the remark this answers is not in the document]
Comment (movito) [reply .github/workflows/sync-core-scripts.yml]: Good catch — fixed in ab071b8. Removed the 'Pull-based alternative' block from the generated PR body. `project sync` lands in PR 2; PR 1 must not point consumers at a command they don't have yet. (The KIT-ADR-0026 reference at the footer stays — that's design provenance, not an instruction.)

File: scripts/core/sync_from_manifest.py [hunk 13]
```diff
@@ -0,0 +1,759 @@
...
+        status = "drift" if should_write else "clean"
+    else:
+        if should_write:
+            _apply(plan, new_manifest, target_root)
+        if warnings:
+            status = "applied_with_warnings"
+        elif content_changed:
+            status = "applied"
+        else:
+            status = "clean"
```
[the remark this answers is not in the document]
Comment (movito) [reply scripts/core/sync_from_manifest.py]: Fixed in ab071b8. A run that writes nothing but produces warnings (e.g. a manifest entry whose source is missing) now reports a distinct `warnings` status (still exit 1) instead of `applied_with_warnings` — the old label claimed a write that never happened. Added `test_warnings_without_write_do_not_claim_applied`.

File: tests/test_sync_from_manifest.py [hunk 14]
```diff
@@ -0,0 +1,663 @@
...
+                "--source",
+                str(missing),
+                "--target",
+                str(target),
+            ],
+            capture_output=True,
+            text=True,
+            timeout=30,
+        )
+        assert manifest_err.returncode == 3, manifest_err.stderr
```
[the remark this answers is not in the document]
Comment (movito) [reply tests/test_sync_from_manifest.py:532]: Fixed in e6f5392 — added explicit `check=False` to all three `subprocess.run` calls. Agreed the CWE-78 flag is a false positive (args are local pytest `Path` fixtures); `check=False` makes the intentional non-zero-exit assertions explicit.

File: .github/workflows/sync-core-scripts.yml [hunk 15]
```diff
@@ -49,147 +70,84 @@ jobs:
 
       - name: Sync from manifest
         id: sync
+        if: >-
+          github.event_name == 'push'
+          || github.event.inputs.repo == ''
+          || github.event.inputs.repo == matrix.repo
```
[the remark this answers is not in the document]
Comment (movito) [reply .github/workflows/sync-core-scripts.yml:102]: Fixed in 1e0c251. Added a `validate-dispatch` job that runs only for workflow_dispatch with a non-empty `repo`, rejects any value outside the known repo list (exit 1 + `::error::`), and gates `sync` via `needs` + an `always()` guard so push runs and valid dispatches still proceed while a typo'd repo now fails the run loudly. Input passed via env (not interpolated) to avoid injection.

**Comment by movito** [comment]
**Note for the KIT-0036 feature-developer (PR 2, `feature/KIT-0036-project-sync`):**

PR #62 is **merged to `main`** (squash commit `0124c18`, 2026-07-04). `docs/DISTRIBUTION-ARCHITECTURE.md` v1.0.0 is now on `main` — your D8 blocker is cleared. One extra commit landed on the PR before merge: a `text` language tag on the plain-text diagram fence (CodeRabbit markdownlint nit), nothing structural.

As you planned: merge `main` into your branch, then apply the v1.1.0 bump. While you're in there, two content notes for v1.1.0 from the planning session:
1. The **'In transition'** section and the **'agents do not auto-propagate'** TL;DR bullet describe the pre-KIT-ADR-0026 world — the pull channel you're shipping changes Channel B's story; update both alongside the new `project sync` section.
2. The **'keeping everything updated' loop** section should gain the pull path (`./scripts/core/project sync`) next to the push-PR path.

— posted on behalf of the operator from the planning session that authored KIT-ADR-0026.
